XWIKI-22667: Introduce CSS equivalents for all LESS variables

Sereza7 opened the pull request:
# Jira URL

<!-- Add the link to the corresponding JIRA issue referenced in a commit message. Unless this is a [Misc] commit,
see https://dev.xwiki.org/xwiki/bin/view/Community/DevelopmentPractices#HRule:Don27tcreateunnecessaryissues
-->

https://jira.xwiki.org/browse/XWIKI-22667

# Changes

## Description

<!-- Describe the main changes brought in this PR. -->

* Added a bunch of CSS variables to make it possible to move some stylesheets from LESS to CSS. In this first pack, there's about 120 variables.
* In priority, I wanted to make sure anything set in the color themes was converted nicely.

## Clarifications

<!-- Provide extra hints to make it easier to understand the PR. Those could be:
* Explanation of choices made in this PR
* Anchor towards extra resources needed to understand the context of this PR (e.g., a forum proposal).
* Links to other issues this issue depends on
-->

* This is the first step of a proposal described on https://design.xwiki.org/xwiki/bin/view/Proposal/MovingawayfromLESSCSS
* Those CSS variables are not API content yet (explained it in the template comments), there'll probably be a lot of changes. However, this is a good base and it's quite straightforward to extend or reduce it.
* This is code we don't use actively yet. I can't be bothered to source this statement, but AFAIR we should not add code to XS that's not useful (avoid the mindset "this might be useful one day" that ends up in dead code cluttering the codebase). However it's been discussed and there'll definitely be a lot of uses very soon. IMO it's alright to add it by itself. 

# Screenshots & Video

<!-- If this PR introduces any UI change, it's recommended to highlight it with before/after screenshots 
or even a screen recording for complex interactions. 
-->
No UI change yet!
We're only making things available in CSS:
![image](https://github.com/user-attachments/assets/1b18a915-5929-467c-a61c-96643a93314b)
Note that the generated CSS doesn't look very nice, but it compiles and the source template is legible so it's alright IMO.
![image](https://github.com/user-attachments/assets/4badd7b8-bf99-4d39-a9b1-31ae1bf80c7e)
All variables needed for Iceberg are here. Some lines are in grey, this is because this variable value is overridden by another line later. This is not the most efficient, but it doesn't cost much (browsers optimize this kind of things on their side, it seems Firefox just straight up just disables the lines) and it's a good safety against bugs. E.g. for one reason or another, we decide to remove --list-group-link-hover-color , but we don't update the value of --component-active-color in the Iceberg colorTheme. We still have a base value for --component-active-color instead of nothing at all.

# Executed Tests

<!-- Especially important for regression fixes. 
Indicate how changes were tested (e.g., what maven commands were run to validate them).
-->

Manual tests with Iceberg. Tried out the font-size functions, they work properly (needed a hack and a few tries before that worked, it's some pretty advanced CSS operations ^^'). Tried out a couple colors, they work as expected.
The Firefox devtool highlight all the variables in formulas correctly, this means they are actually working and not just string values in the hiding. 

I did not test all the other colorThemes, the current changes do not break them BUT they will be way more complicated to migrate than Iceberg. Iceberg migration is pretty much done with those changes already. Bootswatch colorThemes for example have massive LESS sheets in their `@lessCode` parameter. Those sheets still need to be migrated/duplicated to apply their changes to CSS variables, it doesn't just work from scratch with what I've put on here. 

I didn't write any automatic test yet, I'd rather check in some Flamingo test once I updated a Flamingo stylesheet to use CSS variables. Not sure how we'd check CSS variables in the Java test suite anyways.

# Expected merging strategy

* Prefers squash: Yes <!-- No — Explain why. -->
* Backport on branches:
  * None, this is a rather large change, will be followed by large changes and should only be on 17.X
___
# Points to clarify before a merge
* Are the name and type of the file correct? I tried ending it in `.css.vm` but it wouldn't work as a stylesheet `link`. Right now the code completion and linting are horrid but I'm not sure how we could improve that.
* Do we agree on the priority of the variable sets? XSspecific>Bootstrap>oldcolortheme

manuelleduc commented on xwiki-platform-core/xwiki-platform-flamingo/xwiki-platform-flamingo-skin/xwiki-platform-flamingo-skin-resources/src/main/resources/flamingo/cssVariablesInit.css:
```diff
@@ -0,0 +1,309 @@
+"breadcrumb-bg": "color",
+"breadcrumb-color": "color",
+"breadcrumb-link-color": "color",
+"breadcrumb-separator": "escapedText",
```
I see `escapedText` being used later in the code. But does the other values have associated semantics?

Sereza7 replied:
I didn't use the other values yet. Down the line we can use the `@property` annotation, but for now I just went with the "easy" one line instantiation.

manuelleduc commented on xwiki-platform-core/xwiki-platform-flamingo/xwiki-platform-flamingo-skin/xwiki-platform-flamingo-skin-resources/src/main/resources/flamingo/cssVariablesInit.css:
```diff
@@ -0,0 +1,309 @@
+"100vw": "100vw",
+"int-viewport-width": "calc(10000 * tan(atan2(var(--100vw), 10000px)))",
+"font-size-document-title": "calc(max(var(--min-document-title-font-size) * 1px, min(var(--max-document-title-font-size) * 1px, var(--min-document-title-font-size) * 1px + (var(--max-document-title-font-size) - var(--min-document-title-font-size)) * (var(--int-viewport-width) - var(--min-screen-size)) / (var(--max-screen-size) - var(--min-screen-size)))))",
+"font-size-h1": "calc(max(var(--min-h1-font-size), min(var(--max-h1-font-size), var(--min-h1-font-size) + (var(--max-h1-font-size) - var(--min-h1-font-size)) * (var(--int-viewport-width) - var(--min-screen-size)) / (var(--max-screen-size) - var(--min-screen-size)))))",
```
I'm not fan of seeing https://github.com/xwiki/xwiki-platform/blob/92c46a63a954990fb64e95b78a4a4e112cb1583a/xwiki-platform-core/xwiki-platform-flamingo/xwiki-platform-flamingo-skin/xwiki-platform-flamingo-skin-resources/src/main/resources/flamingo/less/type.less#L32 being duplicated.

Sereza7 commented:
@michitux I could factorize the sources of truth for LESS/CSS variables based on the oldcolorthemes and flamingo colorthemes. I added both of the lists containing the shared data  in a new template file used by both the LESS variable generator and the CSS variable generator. 
In addition, I updated the comments to make the link between the leftover lists easier to see. This way, we can avoid creating unwanted discrepancies between the LESS and CSS variable sets later.

michitux commented on xwiki-platform-core/xwiki-platform-flamingo/xwiki-platform-flamingo-skin/xwiki-platform-flamingo-skin-resources/src/main/resources/flamingo/cssVariablesInit.css:
```diff
@@ -0,0 +1,202 @@
+"font-size-h3": "calc(max(var(--min-h3-font-size), min(var(--max-h3-font-size), var(--min-h3-font-size) + (var(--max-h3-font-size) - var(--min-h3-font-size)) * (var(--int-viewport-width) - var(--min-screen-size)) / (var(--max-screen-size) - var(--min-screen-size)))))",
+"font-size-h4": "calc(max(var(--min-h4-font-size), min(var(--max-h4-font-size), var(--min-h4-font-size) + (var(--max-h4-font-size) - var(--min-h4-font-size)) * (var(--int-viewport-width) - var(--min-screen-size)) / (var(--max-screen-size) - var(--min-screen-size)))))",
+"font-size-h5": "var(--font-size-base)",
+"font-size-h6": "calc(max(var(--min-h6-font-size), min(var(--max-h6-font-size), var(--min-h6-font-size) + (var(--max-h6-font-size) - var(--min-h6-font-size)) * (var(--int-viewport-width) - var(--min-screen-size)) / (var(--max-screen-size) - var(--min-screen-size)))))",
```
Did you verify that these font sizes are actually the same as the ones calculated with LESS? Could we add a UI test to verify this, e.g., by putting the same text always twice in a `div` that has the font size once set with a LESS style sheet on the page and once with CSS and then measuring the height of the elements in a JavaScript snippet through Selenium?

Sereza7 replied:
> Did you verify that these font sizes are actually the same as the ones calculated with LESS? 

No however, the operations are working properly and I don't see why they would output a different result given we input the same thing (which we do not, we want to switch the `font-size-base` to em or rem instead of px). The operations here are pretty standard arithmetic and as long as they output something, I assumed the calculations were correct. From my tests on local instances, the output values looked correct.


> Could we add a UI test to verify this, e.g., by putting the same text always twice in a div that has the font size once set with a LESS style sheet on the page and once with CSS and then measuring the height of the elements in a JavaScript snippet through Selenium?

I guess we could test it like this. However IMO this is another case where screenshot based tests would help greatly -- not sure it's worth spending time creating an automated test just for this.

michitux replied:
In my opinion, it is pointless to introduce variables for font sizes if the resulting font size isn't the same as the Less variable. The only use for me of these variables is to be able to give an element the same font size as a heading of that level. If they have a different size, they're not useful to me.

Sereza7 replied:
> The only use for me of these variables is to be able to give an element the same font size as a heading of that level. If they have a different size, they're not useful to me.

Whatever the choice (em/rem), you can use it to set the font size to the same value as the one a heading would have in this context (once we actually do the migration for heading styles) .

The question then is: Do we want heading font sizes to change depending on context or always be the same value.
Example:
Assuming for whatever reason we set a smaller font size for panels. 
1. Do we want the H2s in panels to be the same size as H2 in the content?
OR
2. Do we want the H2s in panels to be smaller than regular H2s, fitting more their context?

___
IMO we might as well go with the option 2 with CSS variables, it's a powerful system that will be easier to extend. I agree that the variable wouldn't be much use in this case, but it doesn't hurt to make it available (we could still remove it even, but it would make some migration a bit less straightforward).

Note that this is just a first draft for CSS variables. As long as we don't implement anything in XS relying on these values it's easy to get back on it. There's a note on the top of this file to make it explicit that those variables can still be subject for change.

@michitux Do you think it'd be useful to open a topic on the forum to discuss this further so that we have a decision made for when we start the migration of headings styles to native CSS and we need a definite answer ?

For now IMO we can go with whatever option, it shouldn't block the whole PR.
Updated the default value to use `rem`s instead of `em`s in https://github.com/xwiki/xwiki-platform/pull/3936/commits/23be8a8f58f365ff8378810bd99f1a3795b5d6e4 :+1:

michitux replied:
> From my tests on local instances, the output values looked correct.

That's part of the "verify" I was asking for. Is this still the case after moving to `rem`?

For me, the important part here is that these heading sizes are actually usable, and part of that is that they are actually the same as the LESS equivalent. I'm not seeing how a screenshot-based test would help here, imho we should just create a UI test that for each heading size from 1 to 6 creates a snippet like

```
(% style="padding: 0; margin: 0" %)
= Heading 1 =

(% style="font-size: var(--font-size-h1); padding: 0; margin: 0" %)
Heading 1
```

and then executes a JavaScript snippet that computes the bounding box of both elements and compares them and ensures that they are sufficiently similar (ideally the same number of pixels, but we need to see if that's possible).

> Note that this is just a first draft for CSS variables. As long as we don't implement anything in XS relying on these values it's easy to get back on it. There's a note on the top of this file to make it explicit that those variables can still be subject for change.

I don't see the point of this PR and of including the CSS in all pages by default if we don't use them. In my opinion, we should use them from now on every time we touch any styles.

Sereza7 replied:
> That's part of the "verify" I was asking for. Is this still the case after moving to rem?

For most of XWiki, rem or em will end up in the same value since we were using an absolute value system. Only the "leaf" elements and the root had their font-sizes defined.

Note: I tested it and there's a difference. Currently bootstrap sets the `font-size` on the body element and not the `:root`. We'll need to change this anyways. When setting the same size as the body on the root, the result looks exactly the same for regular titles (without nesting into another title, which shouldn't happen in real use cases anyways).

Things were looking weird for H6 (10% smaller than before). I did make a mistake on the formula. The scaling factors were reverted in the formulas and I didn't notice it. Fixed in https://github.com/xwiki/xwiki-platform/pull/3936/commits/88e8a181d677c277bc12f78de4be08dd4b646c8d :+1: 

Here is a video demo to showcase the similarities between the current values, the `em` values and the `rem` values. 

https://github.com/user-attachments/assets/38a3752b-6bdc-4218-814e-55d03f0cf3a2

Note that for the `rem` values we need to set the `:root` font-size to be `font-size-base` for it to work properly. The em works directly "as expected" with the value on the `body` element.

Those values are definitely usable. 

> and then executes a JavaScript snippet that computes the bounding box of both elements and compares them and ensures that they are sufficiently similar (ideally the same number of pixels, but we need to see if that's possible).

Okay. I don't fully agree with that but it can be discussed further another day, on the PR where we move the headings styles from LESS to native CSS. IMO adding just one automated test for those values here would be inconsistent. We'd need to test all values and I don't think it's very relevant / useful right now.

 > >  Note that this is just a first draft for CSS variables. As long as we don't implement anything in XS relying on these values it's easy to get back on it. There's a note on the top of this file to make it explicit that those variables can still be subject for change.

> I don't see the point of this PR and of including the CSS in all pages by default if we don't use them. In my opinion, we should use them from now on every time we touch any styles.

IMO this is too big of a change to ship it "on the side" of another UI update. My point is that what we set in this PR will likely need to be changed when we actually use it. It's a skeleton that we can easily update and reach something stable from. I agree that we should use them from now on every time we touch any styles. IMO we don't need to provide a stable API at first, we'll refine our set of variables when using them in XS and once we've transfered enough (possibly all) of XS code to using CSS variables, we can say it's API.

michitux reviewed (changes requested):
Much better overall, I like the deduplication. There are just some details regarding some of the properties where I'm not sure/I think we should change them.

Sereza7 commented on xwiki-platform-core/xwiki-platform-flamingo/xwiki-platform-flamingo-skin/xwiki-platform-flamingo-skin-resources/src/main/resources/flamingo/cssVariablesInit.css:
```diff
@@ -0,0 +1,202 @@
+"font-family-sans-serif": '"Helvetica Neue", Helvetica, Arial, sans-serif',
+"font-family-serif": 'Georgia, "Times New Roman", Times, serif',
+"font-family-monospace": 'Georgia, "Times New Roman", Times, serif',
+"font-size-base": "1em",
```
I guess it depends how we use this variable.

The LESS equivalent is defined in px, so the closest default would be `rem` (which is also the same size wherever in the doc, opposite to `em` that changes depending on context). However, I think it's a better practice to start calculating font-sizes depending on context. IMO ideally we'd just stop using this variable and use directly `<x>em` for sizing most things instead.

Using nested H6 (wrong use case but displays the difference in between both well). CSS and HTML used are found in the first screenshot.
With `1em` as the base value
![image](https://github.com/user-attachments/assets/39678fba-f4f6-439c-8b9a-b2f1bedf6991)
With `1rem` as the base value
![image](https://github.com/user-attachments/assets/7ef10630-a2d8-4273-9d42-5bca59add8fd)

The variable calculations are done at the level of the root element, but depending on the unit of `font-size-base`, they will output the result in different units that are evaluated in the context where they are used.

IMO it makes sense that the nested ones are getting smaller and smaller.

__
If we go with it, we'll need to take care when migrating various font sizes to using CSS variables.

michitux replied:
I'm not convinced that a heading `h2` inside an element of smaller font size should be smaller than in another element. See also my other comment on the heading sizes.

Sereza7 replied:
Started the topic at https://forum.xwiki.org/t/reworking-our-font-size-css-rem-vs-em/16717

Sereza7 commented on xwiki-platform-core/xwiki-platform-flamingo/xwiki-platform-flamingo-skin/xwiki-platform-flamingo-skin-resources/src/main/resources/flamingo/cssVariablesInit.css:
```diff
@@ -0,0 +1,202 @@
+#set($xsVariables = {
+"figure-caption-font-size": "90%",
+"figure-caption-color": "var(--text-muted)",
+"loginFormMaxWidth": "450px",
```
My bad, I removed it in https://github.com/xwiki/xwiki-platform/pull/3936/commits/3a992dff9561cbeccf16566b057a668cc0ea07fc :+1: 

We don't want to carry over inconsistencies from LESS.

michitux reviewed (approved):
Looks good to me in general now, just a few small comments and the question if we should add a UI test for heading sizes.

Sereza7 commented on xwiki-platform-core/xwiki-platform-flamingo/xwiki-platform-flamingo-skin/xwiki-platform-flamingo-skin-resources/src/main/resources/flamingo/variablelist.vm:
```diff
@@ -0,0 +1,155 @@
+## The variables defined in this file were added in XWiki 17.2.0
+## They should not be considered as API yet.
+## After some first "private" uses and some unavoidable cleanup/completion, 
+## this set of CSS variables should become an Unstable API (end be properly documented...).
```
This would be a TODO if we could do it now. IMO we should wait before writing detailed doc for it because it will make it more difficult to change anything in this set of variables. I think it's important to make it API and document it accordingly when we actually settle on things, I added those lines to make sure we don't forget it (and share my vision on it).

Right now LESS variables have no proper documentation at all. There's barely any documentation on styles and classes. I prefer having a note here to avoid making the same mistake :)

Cleaned up a bit the wording in https://github.com/xwiki/xwiki-platform/pull/3936/commits/6a0422d1d950f7098a6250ef2bbb7f16389df05b but the meaning is the same.

Sereza7 replied:
Continuing the discussion on https://github.com/xwiki/xwiki-platform/pull/3936#discussion_r2012122002

manuelleduc commented on xwiki-platform-core/xwiki-platform-flamingo/xwiki-platform-flamingo-skin/xwiki-platform-flamingo-skin-resources/src/main/resources/flamingo/cssVariablesInit.css:
```diff
@@ -0,0 +1,202 @@
+## The variables defined in this file were added in XWiki 17.2.0
+## They should not be considered as API yet.
+## After some first "private" uses and some unavoidable cleanup/completion, 
+## this set of CSS variables should become an Unstable API (end be properly documented...).
```
```suggestion
## this set of CSS variables should become an Unstable API (and be properly documented...).
```
But I'm not sure about this block of code.
I suggest to
1. clearly mark the version where the variables are introduced (17.3.0RC1)
1. possibly explain that they will become stable at the end of the next major cycle (as per our development rule)

What we can't do, sadly is giving a special name to unstable variables.
And there is no way to annotate CSS variables either.

Sereza7 replied:
Updated the version and fixed the typo :+1: 

___
I just wanted to make clear that this isn't even worth calling it an unstable API yet. Personally, I'd expect an unstable API to stay 90% the same or so from the stable state. This is not a formal definition by any means but in practice that's what happens AFAIK and that's where my expectations come from.
Here, I just threw a set of variables together without thinking it too in depth. I'd expect the end version to have barely 50% of the variables with unchanged names and defaults.

I'll remove all mentions of API in those comments and just word it like you did `those variables will become stable at the end of the 17th cycle`.

Done in https://github.com/xwiki/xwiki-platform/pull/3936/commits/9978def0dbb8d6fe8642c76e0668341b3cd74178 :+1:

Sereza7 commented on xwiki-platform-core/xwiki-platform-flamingo/xwiki-platform-flamingo-skin/xwiki-platform-flamingo-skin-resources/src/main/resources/flamingo/variablelist.vm:
```diff
@@ -0,0 +1,155 @@
+## Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+## 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+## ---------------------------------------------------------------------------
+## The variables defined in this file were added in XWiki 17.2.0
```
Addressed in https://github.com/xwiki/xwiki-platform/pull/3936/commits/6a0422d1d950f7098a6250ef2bbb7f16389df05b :+1:

manuelleduc reviewed (changes requested):
Once the issue with `cssVariableInit.css` being badly minified is fixed, I think the PR will be ready to merge.

Sereza7 commented:
![image](https://github.com/user-attachments/assets/4e1b0a4a-6ed7-42a6-880a-1f5bb19f6536)

To test [the latest changes that fixed minification](https://github.com/xwiki/xwiki-platform/pull/3936/commits/d487aa84a16b166479564cfa673b61cc5be63bd4), I build the module with `mvn clean install -f xwiki-platform-core/xwiki-platform-flamingo/xwiki-platform-flamingo-skin/xwiki-platform-flamingo-skin-resources`, then extracted the resulting jar under `webapps/xwiki/skins`. As we can see in the screenshot above:
* by default, the minified version is used
* The minified version contains all the variables as expected.
* Those variables can be used as expected.

Thanks Manuel for spotting this issue with the PR and helping solve it!
___
@manuelleduc The PR should be ready :)

michitux commented on xwiki-platform-core/xwiki-platform-flamingo/xwiki-platform-flamingo-skin/xwiki-platform-flamingo-skin-resources/src/main/resources/flamingo/cssVariablesInit.css:
```diff
@@ -0,0 +1,204 @@
+#end
+}
+#end
+*/$!css
```
This works, minification doesn't remove the whole comment?

manuelleduc replied:
The comment with velocity pattern is inspired from https://github.com/xwiki/xwiki-platform/blob/226762d964f8a4ba298d84a3bacad3eb42efed03/xwiki-platform-core/xwiki-platform-flamingo/xwiki-platform-flamingo-skin/xwiki-platform-flamingo-skin-resources/src/main/resources/flamingo/style.css#L10
afaik the content of the comment is kept during minification.
Apparently, the magic in the exclamation mark at the end of the opening on the comment (`/*!`)
I tried locally, and removing the exclamation mark makes the comment disappear during minification.
Maybe @mflorea knows where this comes from?

mflorea replied:
> Maybe @mflorea knows where this comes from?

@manuelleduc this is most probably needed to avoid YuiCompressor from removing the comment (with the Velocity code) when minifying the CSS file. The alternative is to exclude this CSS file from minification (in the POM).
